src/haz3lcore/statics/Statics.re

```diff
@@ -328,192 +141,188 @@ and uexp_to_info_map =
-      union_m([m_pat, m_def, m_body]),
+      ~self=Info.join(Fun.id, [cons.ty, alt.ty], branch_ids, ctx),
+      ~free=Ctx.union([cond.free, cons.free, alt.free]),
+      m,
     );
   | Match(scrut, rules) =>
-    let (ty_scrut, free_scrut, m_scrut) = go(~mode=Syn, scrut);
-    let (pats, branches) = List.split(rules);
-    let pat_infos =
-      List.map(
-        upat_to_info_map(~is_synswitch=false, ~mode=Typ.Ana(ty_scrut)),
-        pats,
-      );
-    let branch_infos =
-      List.map2(
-        (branch, (_, ctx_pat, _)) =>
-          uexp_to_info_map(~ctx=VarMap.concat(ctx, ctx_pat), ~mode, branch),
-        branches,
-        pat_infos,
-      );
-    let branch_sources =
-      List.map2(
-        (e: Term.UExp.t, (ty, _, _)) => Typ.{id: Term.UExp.rep_id(e), ty},
-        branches,
-        branch_infos,
+    let (scrut, m) = go(~mode=Syn, scrut, m);
+    let (ps, es) = List.split(rules);
+    let branch_ids = List.map(UExp.rep_id, es);
+    let (ps, m) =
+      map_m(go_pat(~is_synswitch=false, ~mode=Typ.Ana(scrut.ty)), ps, m);
+    let p_ctxs = List.map(Info.pat_ctx, ps);
+    let (es, m) =
+      List.fold_left2(
+        ((es, m), e, ctx) =>
+          go'(~ctx, ~mode, e, m) |> (((e, m)) => (es @ [e], m)),
+        ([], m),
+        es,
+        p_ctxs,
       );
-    let pat_ms = List.map(((_, _, m)) => m, pat_infos);
-    let branch_ms = List.map(((_, _, m)) => m, branch_infos);
-    let branch_frees = List.map(((_, free, _)) => free, branch_infos);
-    let self = Typ.Joined(Fun.id, branch_sources);
-    let free = Ctx.union([free_scrut] @ branch_frees);
-    add(~self, ~free, union_m([m_scrut] @ pat_ms @ branch_ms));
+    let e_tys = List.map(Info.exp_ty, es);
+    let e_frees =
+      List.map2(Ctx.free_in(ctx), p_ctxs, List.map(Info.exp_free, es));
+    add(
+      ~self=Info.join(Fun.id, e_tys, branch_ids, ctx),
+      ~free=Ctx.union([scrut.free] @ e_frees),
+      m,
+    );
+  | TyAlias(typat, utyp, body) =>
+    let m = utpat_to_info_map(~ctx, ~ancestors, typat, m) |> snd;
+    switch (typat.term) {
+    | Var(name)
+        when !Form.is_base_typ(name) && Ctx.lookup_alias(ctx, name) == None =>
+      /* NOTE(andrew): This is a slightly dicey piece of logic, debatably
+         errors cancelling out. Right now, to_typ returns Unknown(TypeHole)
+         for any type variable reference not in its ctx. So any free variables
+         in the definition would be oblierated. But we need to check for free
+         variables to decide whether to make a recursive type or not. So we
+         tentatively add an abtract type to the ctx, representing the
+         speculative rec parameter. */
+      let (ty_def, ctx_def, ctx_body) = {
+        let ty_pre =
+          UTyp.to_typ(Ctx.add_abstract(ctx, name, Id.invalid), utyp);
+        switch (utyp.term) {
+        | USum(_) when List.mem(name, Typ.free_vars(ty_pre)) =>
+          let ty_rec = Typ.Rec("α", Typ.subst(Var("α"), name, ty_pre));
```

Comment:
why can't we use the name instead of α?

Reply:
Error prone; easy to make substitution-related implementation errors which are hard to track. AFAICT the only disadvantage is CI, and this should be dealt with at another level of abstraction